src/yz_solrq.erl

```diff
@@ -0,0 +1,817 @@
+    gen_server:cast(QPid, {batch_complete, Index, Message}).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+%% @private
+init([]) ->
+    {ok, read_appenv(#state{})} .
+
+handle_call({index, Index, E}, From, State) ->
+    ?PULSE_DEBUG("index.  State: ~p~n", [debug_state(State)]),
+    State2 = inc_qlen_and_maybe_unblock_vnode(From, State),
+    IndexQ = enqueue(E, get_indexq(Index, State2)),
+    IndexQ2 = maybe_request_worker(Index, IndexQ),
+    IndexQ3 = maybe_start_timer(Index, IndexQ2),
+    NewState = update_indexq(Index, IndexQ3, State2),
+    ?PULSE_DEBUG("index.  NewState: ~p~n", [debug_state(NewState)]),
+    {noreply, NewState};
+handle_call(status, _From, #state{} = State) ->
+    {reply, internal_status(State), State};
+handle_call({set_hwm, NewHWM}, _From, #state{queue_hwm = OldHWM} = State) ->
+    {reply, {ok, OldHWM}, maybe_unblock_vnodes(State#state{queue_hwm = NewHWM})};
+handle_call(get_hwm, _From, #state{queue_hwm = HWM} = State) ->
+    {reply, HWM, State};
+handle_call({set_index, Index, Min, Max, DelayMS}, _From, State) ->
+    IndexQ = get_indexq(Index, State),
+    IndexQ2 = maybe_request_worker(Index,
+                                   IndexQ#indexq{batch_min = Min,
+                                                 batch_max = Max,
+                                                 delayms_max = DelayMS}),
+    OldParams = {IndexQ#indexq.batch_min,
+                 IndexQ#indexq.batch_max,
+                 IndexQ#indexq.delayms_max},
+    {reply, {ok, OldParams}, update_indexq(Index, IndexQ2, State)};
+handle_call({set_purge_strategy, NewPurgeStrategy},
+    _From,
+    #state{purge_strategy=OldPurgeStrategy} = State) ->
+    {reply, {ok, OldPurgeStrategy}, State#state{purge_strategy=NewPurgeStrategy}};
+handle_call(cancel_drain, _From, State) ->
+    {noreply, NewState} = handle_cast(drain_complete, State),
+    {reply, ok, NewState};
+handle_call(all_queue_len, _From, #state{all_queue_len=Len} = State) ->
+    {reply, Len, State};
+handle_call(reload_appenv, _From, State) ->
+    {reply, ok, read_appenv(State)}.
+handle_cast({request_batch, Index, HPid}, State) ->
+    State2 = send_entries(HPid, Index, State),
+    {noreply, State2};
+
+%%
+%% @doc Handle the drain message.
+%%
+%%      The drain message is sent via the drain/1 function
+%%      in this module, which is called by the solrq_drain_fsm
+%%      during its prepare state, typically as the result of
+%%      a request to drain the queues.
+%%
+%%      This handler will iterate over all IndexQs in the
+%%      solrq, and initiate a drain on the queue, if it is currently
+%%      non-empty.
+%% @end
+%%
+handle_cast({drain, DPid, Token, Partition}, #state{indexqs = IndexQs} = State) ->
+    ?PULSE_DEBUG("drain{~p=DPid, ~p=Token, ~p=Partition}.  State: ~p~n", [debug_state(State), DPid, Token, Partition]),
```

Comment:
This module needs to be wrapped to 80 or 120 columns, or whatever our coding conventions dictate.